### PR TITLE
receive: fail early if ketama hashring is configured with number of nodes lower than the replication factor

### PR DESCRIPTION
**Pull request by MichaHoffmann**
Addresses #5639

* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

Construction of the ketama hashring can busyloop forever if less endpoints than replication factor was configured even briefly.

## Verification

Added a unittest.

### CODE REVIEW

File: pkg/receive/hashring.go [hunk 1]
```diff
@@ -276,7 +285,11 @@ func HashringFromConfigWatcher(ctx context.Context, algorithm HashringAlgorithm,
 			if !ok {
 				return errors.New("hashring config watcher stopped unexpectedly")
 			}
-			updates <- newMultiHashring(algorithm, replicationFactor, cfg)
+			h, err := newMultiHashring(algorithm, replicationFactor, cfg)
+			if !ok {
```
Comment (matej-g) [inline pkg/receive/hashring.go:289]: You probably meant this right?

```suggestion
			if err != nil {
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -15,6 +15,7 @@ We use *breaking :warning:* to mark changes that are not backward compatible (re
 ### Fixed
 
 ### Changed
+- [#6168](https://github.com/thanos-io/thanos/pull/6168) Receiver: make ketama hashring fail early when configured in a bad way.
```
Comment (matej-g) [inline CHANGELOG.md:18]: Nit: I'd specify what exactly is the change here, e.g. something like
```suggestion
- [#6168](https://github.com/thanos-io/thanos/pull/6168) Receiver: Make ketama hashring fail early when configured with number of nodes lower than the replication factor.
```

**Review by matej-g — COMMENTED** [review]
Thanks @MichaHoffmann, this is looking nice!

I'm not sure why CircleCI is on strike again, I often see the run fails for first-time contributors, I'll try to see if we can fix it.

File: pkg/receive/hashring.go (hunk 1 above)
Comment (mhoffm-aiven) [reply pkg/receive/hashring.go:289]: Oh yeah! Thanks

Comment (MichaHoffmann) [reply pkg/receive/hashring.go:289]: :heavy_check_mark:

File: CHANGELOG.md (hunk 2 above)
Comment (MichaHoffmann) [reply CHANGELOG.md]: Thanks! I amended my commit with your suggestion!

File: pkg/receive/handler_test.go [hunk 3]
```diff
@@ -576,7 +579,10 @@ func testReceiveQuorum(t *testing.T, hashringAlgo HashringAlgorithm, withConsist
 		},
 	} {
 		t.Run(tc.name, func(t *testing.T) {
-			handlers, hashring := newTestHandlerHashring(tc.appendables, tc.replicationFactor, hashringAlgo)
+			handlers, hashring, err := newTestHandlerHashring(tc.appendables, tc.replicationFactor, hashringAlgo)
+			if err != nil {
```
Comment (fpetkovski) [inline pkg/receive/handler_test.go:583]: We usually use `testutil.Ok`.

File: pkg/receive/hashring.go [hunk 4]
```diff
@@ -110,9 +110,14 @@ type ketamaHashring struct {
 	numEndpoints uint64
 }
 
-func newKetamaHashring(endpoints []string, sectionsPerNode int, replicationFactor uint64) *ketamaHashring {
+func newKetamaHashring(endpoints []string, sectionsPerNode int, replicationFactor uint64) (*ketamaHashring, error) {
 	numSections := len(endpoints) * sectionsPerNode
 
+	if len(endpoints) < int(replicationFactor) {
+		return nil, fmt.Errorf("ketama: amount of endpoints needs to be larger than replication factor")
```
Comment (fpetkovski) [inline pkg/receive/hashring.go:117]: We prefer using `errors.New` since it includes a stacktrace when printed out.

**Review by fpetkovski — COMMENTED** [review]
Thanks, looks good overall. Just some small nits about errors and assertions.

File: pkg/receive/handler_test.go (hunk 3 above)
Comment (mhoffm-aiven) [reply pkg/receive/handler_test.go:583]: I chose that to be consistent with the other assertions in this test. Its probably best to update all of them. While i'm at it should i also convert them to `testutil.Ok` form?

File: pkg/receive/hashring.go (hunk 4 above)
Comment (mhoffm-aiven) [reply pkg/receive/hashring.go:117]: Good point, that was just out of habit; ill convert it after work!

File: pkg/receive/handler_test.go (hunk 3 above)
Comment (fpetkovski) [reply pkg/receive/handler_test.go:583]: I think that's a good point, let's then leave it as it is for consistency.

File: pkg/receive/hashring.go (hunk 4 above)
Comment (MichaHoffmann) [reply pkg/receive/hashring.go]: changed it to `errors.New`

**Review by matej-g — APPROVED** [review]
Nice work @MichaHoffmann 💪

**Review by fpetkovski — APPROVED** [review]
Thanks!

**Pull request merged**